lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -0,0 +1,427 @@
+        op->setAttr("layout", updatedLayoutAttr);
+      }
+
+      if ((tensorBufferTypeChanged || tensorMemoryLayoutChanged) &&
+          op->getAttrDictionary().get("memory_config")) {
+        // Create the output memory config attribute
+        ttnn::MemoryConfigAttr updatedMemConfigAttr =
+            ttnn::MemoryConfigAttr::get(
+                rewriter.getContext(),
+                ttnn::TensorMemoryLayoutAttr::get(rewriter.getContext(),
+                                                  tensorMemoryLayout),
+                ttnn::BufferTypeAttr::get(rewriter.getContext(),
+                                          tensorBufferType),
+                ttnn::ShardSpecAttr::get(
+                    op.getContext(),
+                    ttnn::ShapeAttr::get(rewriter.getContext(),
+                                         outputLayout.getMemref().getShape())));
+        op->setAttr("memory_config", updatedMemConfigAttr);
+      }
+    });
+
+    // Propagate output change into next DPS inits operands that uses this
+    // result
+    propagateOutputChangesIntoDPSInits(outputResult, rewriter);
+  }
+
+  return modified;
+}
+
+// Rewriter to apply workarounds to the operands of TTNN operations.
+// This rewriter applies the workarounds to the input and output operands of
+// TTNN operations.
+class TTNNOperandsWorkaroundsRewriter
+    : public OpInterfaceRewritePattern<wa::TTNNWorkaroundInterface> {
+public:
+  TTNNOperandsWorkaroundsRewriter(MLIRContext *ctx)
+      : OpInterfaceRewritePattern<wa::TTNNWorkaroundInterface>(ctx) {}
+
+  LogicalResult matchAndRewrite(wa::TTNNWorkaroundInterface op,
+                                PatternRewriter &rewriter) const final {
+
+    // To layout op is a special case, we don't want to rewrite it
+    if (mlir::isa<ttnn::ToLayoutOp>(op.getOperation())) {
+      return failure();
+    }
+
+    // Get the operands workarounds for the current operation
+    wa::TTNNOperandsWorkarounds operandsWorkarounds =
+        op.getOperandsWorkarounds();
+
+    // Apply input workarounds only for tensor operands
+    bool modifiedOperands = false;
+    int inputOperandIndex = 0;
+    for (size_t i = 0;
+         i < operandsWorkarounds.getInputOperandWorkarounds().size(); i++) {
+      wa::TTNNOperandWorkarounds inputWorkaround =
+          operandsWorkarounds.getInputOperandWorkarounds()[i];
+
+      // No input operand workarounds to apply, hence continue
+      if (!inputWorkaround.hasAnyWorkaround()) {
+        inputOperandIndex++;
+        continue;
+      }
+
+      // Get the next tensor opearand
+      while (!mlir::isa<RankedTensorType>(
+          op->getOperand(inputOperandIndex).getType())) {
+        inputOperandIndex++;
+      }
+
+      OpOperand &inputOperand = op->getOpOperand(inputOperandIndex++);
+      // Apply all workaround changes to the input operand
+      modifiedOperands |=
+          workaroundInputOperand(inputOperand, inputWorkaround, rewriter, op);
+    }
```

Comment:
This whole loop is hard to read with `inputOperandIndex` and `continue`, if I understand this correctly it can be rewritten to be much cleaner (didn't try to compile, so might need some additions). Same for the outputs.
```
auto inputTensors = llvm::make_filter_range(op.getOperands(), isTensorType);
bool modifedOperads = false;
llvm::for_each(llvm::zip_equal(inputTensors, operandsWorkarounds.getInputOperandWorkarounds()), // if there is assumption that number of input tensors and input workarounds has to be the same
    [&](auto p) { modifiedOperands |= workaroundInputOperand(std::get<0>(p), std::get<1>(p), rewriter, op); } 
```